esx_service/vmdk_ops.py

```diff
@@ -252,31 +253,62 @@ def getVMDKUuid(vmdk_path):
     except:
         return None
 
+# Create and return the devFS path for a VSAN object UUID.
+# Also returns true if clean up is needed.
+# cleanup_vsan_devfs should be called to clean up before
+# removing the VSAN object.
+def get_vsan_devfs_path(uuid):
+
+    logging.debug("Got volume UUID %s", uuid)
+
+    # Objtool creates a link thats usable to 
+    # read write to vsan object.
+    cmd = "{0} {1}".format(OBJ_TOOL_CMD, uuid)
+    rc, out = RunCommand(cmd)
+    if rc == 0 and os.path.isfile(fpath):
+        return "/vmfs/devices/vsan/{0}".format(uuid), True
+    return None, False
+
+# Clean up vsan devfs path
+def cleanup_vsan_devfs_path(devfs_path)
+    try:
+        os.remove(backing)
+        logging.debug("Unlinked %s for flag %s", backing, cleanup_device)
+	return True
+    except OSError as ex:
+        logging.error("Failed to remove backing device %s",
+                      backing, str(ex))
+    return False
+
+# Returns the flat file for a VMDK.
+def get_backing_flat_file(vmdk_path):
+    return vmdk_path.replace(".vmdk", "-flat.vmdk")
 
```

Comment:
This won't work once snaps are supported (which is pretty soon). Instead please use the vmdk descriptor file to get the backing file. For a delta disk there is no "-flat" in the file name. A delta disk looks like this:
Ubuntu-14.04-64-LTS-IMG-000001-delta.vmdk

And the corresponding descriptor file will clearly say whats the backing, lets use that vs. doing a string replace (was ok till now, not further).

[root@localhost:/vmfs/volumes/523c14b7-e5561042-5e2f-002481aaea6c/Ubuntu-14.04-64-LTS-IMG] cat Ubuntu-14.04-64-LTS-IMG-000001.vmdk
# Disk DescriptorFile
version=1
encoding="UTF-8"
CID=764b5bfd
parentCID=91a8e92f
isNativeSnapshot="no"
createType="vmfsSparse"
parentFileNameHint="/vmfs/volumes/523c14b7-e5561042-5e2f-002481aaea6c/Ubuntu-14.04-64-LTS-IMG/Ubuntu-14.04-64-LTS-IMG.vmdk"
# Extent description
RW 67108864 VMFSSPARSE "Ubuntu-14.04-64-LTS-IMG-000001-delta.vmdk"

# The Disk Data Base
#DDB

ddb.longContentID = "caa378fba2e6a2f300876aa2764b5bfd"


Use the extent file mentioned in the descriptor.